File: data-structures-mastery/01_memory_and_addresses/c/test_memory_and_addresses.c
#include "memory_and_addresses.h"
#include <assert.h>
#include <stdio.h>

int main(void) {
    /* swap_ints must actually exchange the two values */
    int a = 1, b = 2;
    swap_ints(&a, &b);
    assert(a == 2 && b == 1);

    /* array_stride_bytes must equal sizeof(int) for a plain int array --
     * this is the proof that array elements sit back-to-back in memory */
    int arr[3] = {0, 0, 0};
    assert(array_stride_bytes(arr) == (ptrdiff_t)sizeof(int));

    /* swapping a value with itself must be a safe no-op */
    int c = 42;
    swap_ints(&c, &c);
    assert(c == 42);

    printf("All C tests passed.\n");
    return 0;
}

File: data-structures-mastery/01_memory_and_addresses/python/test_memory_and_addresses.py
from memory_and_addresses import swap_via_temp


def test_swap_via_temp_swaps_values():
    arr = [1, 2, 3]
    swap_via_temp(arr, 0, 2)
    assert arr == [3, 2, 1]


def test_swap_via_temp_self_swap_is_noop():
    arr = [7]
    swap_via_temp(arr, 0, 0)
    assert arr == [7]


def test_aliasing_shares_identity():
    a = [1, 2, 3]
    b = a
    assert id(a) == id(b)
    b.append(4)
    assert a == [1, 2, 3, 4]  # mutation through b is visible through a


def test_reassignment_breaks_aliasing():
    a = [1, 2, 3]
    b = a
    b = [9, 9, 9]  # rebinds b; a is untouched
    assert a == [1, 2, 3]
    assert id(a) != id(b)


if __name__ == "__main__":
    test_swap_via_temp_swaps_values()
    test_swap_via_temp_self_swap_is_noop()
    test_aliasing_shares_identity()
    test_reassignment_breaks_aliasing()
    print("All Python tests passed.")
